### DISCUSSION TITLE
DirectML Neural Upscaler (APIs available)

### DISCUSSION THREAD

**Discussion by RyzenAdvanced** [Blinue/Magpie#153] [Ideas]
Dear Developer,

As now AMD RDNA 2 GPUs has proven 4x DirectML performance, maybe you could add DirectML based upscaling API as well?

**DirectML APIs:**
https://mspoweruser.com/microsoft-announce-the-public-release-of-directml-as-a-standalone-api/amp/?__twitter_impression=true

**Example:**
https://46c4ts1tskv22sdav81j9c69-wpengine.netdna-ssl.com/wp-content/uploads/prod/sites/3/2021/01/super_resolution_sample.jpg

**Source about DirectML in RDNA 2 GPUs:**
https://www.google.com/amp/s/wccftech.com/amd-microsoft-bring-tensorflow-directml-to-life-4x-improvement-with-rdna-2-gpus/amp/

Kindly let us know,

Thank you!!

**Comment by RyzenAdvanced** [comment]
@Blinue hopefully,  hear your thoughts on these idea, as this could really make Magpie competitive with deep learning upscale such as DLSS (I think). Very excited to see if you could try add these one :)

**Comment by Blinue** [comment]
Thanks for the suggestion. I tried the official examples and the machine learning-based scaling algorithm was so slow as to be almost unusable on my computer. I only have a NVIDIA card.

There are many reasons preventing me from using this technique.

1. Its performance is unacceptable, at least on my computer. DirectML may be much faster on AMD card, but I have no way to verify it. 

2. It is based on DirectX12, and Magpie currently uses DirectX11. There are far fewer computers that support DirectX12 than DirectX11.

3. DirectML is only applicable to machine learning-based algorithms, and is not usable for traditional algorithms such as FSR.

4. No public machine learning-based algorithm is based on DirectML  at the moment. I believe porting them can be quite difficult.

**Reply by RyzenAdvanced** [reply]
Dear @Blinue,

I understand,  but since now AMD and Microsoft released new drivers that boost x4 DirecrML on RDNA 2 GPUs, is there a chance we could see experimental version of Magpie that uses DirectML to improve scaling?

Thank you!

**Reply by Blinue** [reply]
I can't code for hardware I don't have.

**Comment by RyzenAdvanced** [comment]
Dear @Blinue ,

How about FFT upscsler? 
https://github.com/DTolm/VkResample

Thanks !

**Reply by Blinue** [reply]
FSR is good enough for 3D games.

**Reply by RyzenAdvanced** [reply]
@Blinue understood :)

Does current Magpie uses AMD's denoiser for FSR?
https://github.com/GPUOpen-LibrariesAndSDKs/RadeonImageFilter/tree/master/models

https://gpuopen.com/radeon-image-filtering-library/#benefits

"Supports AMD GPUs as well as those of other vendors using OpenCL™ for AI-accelerated denoising and standard GPU-accelerated filters"

**Reply by RyzenAdvanced** [reply]
> FSR is good enough for 3D games.

Hi @Blinue ,

Are we using AMD Denoiser with FSR?
https://gpuopen.com/fidelityfx-denoiser/

and:
https://gpuopen.com/radeon-image-filtering-library/

**Also, is this one upscaler is something we can test as overlay/injector?**
https://github.com/HasnainRaz/Fast-SRGAN

**Reply by hooke007** [reply]
AMD's denoiser seems to be another one based on dx12.

https://github.com/HasnainRaz/Fast-SRGAN#speed-benchmarks
Too slow for realtime-using.
